openwpm_utils/s3.py

```diff
@@ -29,12 +32,20 @@ def __init__(self, spark_context, s3_directory,
         self._s3_directory = s3_directory
         self._spark_context = spark_context
         self._sql_context = SQLContext(spark_context)
-        self._s3_table_loc = "s3a://%s/%s/visits/%%s/" % (
-            s3_bucket, s3_directory)
-        self._s3_content_loc = "s3a://%s/%s/content/%%s.gz" % (
-            s3_bucket, s3_directory)
-
-    def read_table(self, table_name, columns=None):
+        self._s3_table_loc = "s3a://%s/%s/visits/%%s/" % (s3_bucket, s3_directory)
+        self._s3_content_loc = "s3a://%s/%s/content/%%s.gz" % (s3_bucket, s3_directory)
+        self._incomplete_visit_ids = self.read_table(
+            "incomplete_visits", mode="all"
+        ).select("visit_id")
+        crawl_history = self.read_table("crawl_history", mode="all")
+        self._failed_visit_ids = (
```

Comment:
Since this operation will run for every loaded dataset we should try to avoid using udfs. The underlying checks are pretty simple, so I think we can convert them to pyspark built-in functions. Here's a rough sketch (untested):

```python
crawl_history.groupBy("visit_id").agg(
    F.when(
        F.array_contains("command_status", "critical"), "critical"
    ).when(
        F.array_contains("command_status", "error"), "error"
    ).when(
        F.array_contains("command_status", "neterror"), "neterror"
    ).when(
        F.array_contains("command_status", "timeout"), "timeout"
    ).otherwise(
        "ok"
    ).alias("worst_status")
)
```

see:
* https://spark.apache.org/docs/latest/api/python/pyspark.sql.html#pyspark.sql.functions.array_contains
* https://spark.apache.org/docs/latest/api/python/pyspark.sql.html#pyspark.sql.functions.when

Reply:
Done